# Tratamento dos Dados 🪢

<pre>A base de indicadores criminais por município e ano do Estado do Rio Grande do Sul é publicada no formato Excel. Por meio de raspagem de dados, obtivemos os dados referentes aos anos de 2012 até a presente data.</pre>

<pre>Execute as duas células abaixo para condensar todos os anos da base de indicadores criminais por município do Estado do Rio Grande do Sul em um único DataFrame, de nome <b>df_crimes</b>.</pre>

In [1]:
!pip install --upgrade openpyxl --quiet
!pip install --upgrade pandas --quiet

In [2]:
import pandas as pd

dfs = []

xlsx = pd.ExcelFile('https://raw.githubusercontent.com/heliomacedofilho/idp_fundamentos_2026/refs/heads/main/day%2003/Parte%2001/exercises/Nivel%203/data/crimes_rs_municipios.xlsx')

anos = [year for year in xlsx.sheet_names if all(c.isdigit() for c in year)]

for ano in anos:
    df = pd.read_excel(xlsx, sheet_name=ano)
    df['Ano'] = ano
    dfs.append(df)

df_crimes = pd.concat(dfs)
df_crimes.head()

,Municípios,Homicídio Doloso,Latrocínio,Furtos,Abigeato*,Furto de Veículo,Roubos,Roubo de Veículo,Estelionato,Delitos Relacionados à Armas e Munições,Entorpecentes - Posse,Entorpecentes - Tráfico,Vítimas de Lesão Corp. Seg. Morte,Total/Ano,Ano
0,ACEGUA,0,0,20,7,0,0,0,7,1,0,0,0,35,2024
1,AGUA SANTA,0,0,6,2,1,0,1,11,6,0,1,0,28,2024
2,AGUDO,1,0,42,1,2,0,0,52,2,5,8,0,113,2024
3,AJURICABA,0,0,17,0,1,0,0,20,2,3,0,0,43,2024
4,ALECRIM,0,0,23,3,2,0,0,12,9,2,3,0,54,2024


## Pivoteamento 🥈

### Pivoteamento do Formato Longo em Formato Largo

<pre>A base de dados possui a quantidade de crimes registrada para cada ano e município do Rio Grande do Sul.</pre>

<pre>Crie um novo DataFrame, de nome <b>df_homicidios</b>, contendo somente os valores das colunas <b>Municípios</b>, <b>Homicídio Doloso</b> e <b>Ano</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios = df_crimes[['Municípios', 'Homicídio Doloso', 'Ano']]
```

<br/>

</details>

<pre>Agora, aplique o pivoteamento do formato longo em formato largo, para que
    - o nome do município seja o índice das linhas; e
    - cada coluna contenha os números de homicídio dolosos cometido no ano correspondente.</pre>

<pre>Isto é, modificaremos o DataFrame <b>df_homicidios</b> do formato longo:</pre>

<table>
    <tr>
        <th>Municípios</th>
        <th>Ano</th>
        <th>Homicídio Doloso</th>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2012</td>
        <td>0</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2013</td>
        <td>1</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2014</td>
        <td>0</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2015</td>
        <td>0</td>
    </tr>
    <tr>
        <th>...</th>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2021</td>
        <td>1</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2022</td>
        <td>1</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2023</td>
        <td>0</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2024</td>
        <td>0</td>
    </tr>
    <tr>
        <th>AGUA SANTA</th>
        <td>...</td>
        <td>...</td>
    </tr>
</table>

<pre>Para o formato largo:</pre>

<table>
    <tr>
        <th>Ano</th>
        <th>2012</th>
        <th>2013</th>
        <th>2014</th>
        <th>2015</th>
        <th>2016</th>
        <th>2017</th>
        <th>2018</th>
        <th>2019</th>
        <th>2020</th>
        <th>2021</th>
        <th>2022</th>
        <th>2023</th>
        <th>2024</th>
    </tr>
    <tr>
        <th>Municípios</th>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>0</td>
        <td>1</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>0</td>
        <td>1</td>
        <td>0</td>
        <td>1</td>
        <td>1</td>
        <td>1</td>
        <td>0</td>
        <td>0</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
    </tr>
</table>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios = df_homicidios.pivot(index=['Municípios'], columns='Ano', values='Homicídio Doloso')
```

<br/>

</details>

<pre>Acrescente uma coluna denominada <b>Total/Município</b> no DataFrame <b>df_homicidios</b> com o total (a soma) de homicídios para cada município.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios['Total/Município'] = df_homicidios.sum(axis=1)
```

<br/>

</details>

<pre>Agora, crie um DataFrame <b>df_homicidios_rank</b> com os <b>10</b> municípios do Rio Grande do Sul com maior quantidade de homicídios dolosos, de acordo com a coluna <b>Total/Município</b>. Posteriormente, ordene a tabela resultante pelo nome do município, em ordem alfabética.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios_rank = df_homicidios.sort_values('Total/Município', ascending=False).head(10).sort_index()
```

<br/>

</details>

<pre>Crie uma linha de índice <b>Total/Ano</b> que totalize (some) os valores de cada uma das colunas no DataFrame <b>df_homicidios_rank</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios_rank.loc['Total/Ano'] = df_homicidios_rank.sum(axis=0)
```

<br/>

</details>

<pre>Agora, calcule o valor percentual de cada um dos municípios em relação ao total calculado acima, i.e. utilizando a linha <b>Total/Ano</b> como base.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios_rank = df_homicidios_rank.div(df_homicidios_rank.loc['Total/Ano'])
```

<br/>

</details>

<pre>Por fim, apresente todos os valores do DataFrame <b>df_homicidios_rank</b> com valores percentuais e precisão de duas casas decimais.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_homicidios_rank.map(lambda x: f'{x:.2%}')
```

<br/>

</details>

### Pivoteamento do Formato Largo em Formato Longo

<pre>A base de população estimada do IBGE é publicada em vários formatos no <a href="https://sidra.ibge.gov.br/pesquisa/estimapop/tabelas">Sistema IBGE de Recuperação Automática</a>. Por meio de raspagem de dados, obtivemos os dados referentes aos municípidos do Estado do Rio Grande do Sul, dos anos de 2001 até a presente data.</pre>

<pre>Execute o código abaixo para carregar a base de população estimada do IBGE, filtrada pelos dados referentes aos municípidos do Estado do Rio Grande do Sul, dos anos de 2001 até a presente data, no DataFrame <b>df_populacao</b>.</pre>

In [3]:
df_populacao = pd.read_excel('https://raw.githubusercontent.com/heliomacedofilho/idp_fundamentos_2026/refs/heads/main/day%2003/Parte%2001/exercises/Nivel%203/data/populacao_rs_municipios.xlsx')
df_populacao.head()

,Município,2001,2002,2003,2004,2005,2006,2008,2009,2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021
0,ACEGUA,3970,4002,4036,4109,4149,4188,4312,4347,4430,4465,4638,4671,4702,4731,4759,4858,4901,4942,4981
1,AGUA SANTA,3826,3852,3880,3940,3973,4006,3638,3619,3717,3712,3838,3839,3840,3841,3841,3753,3748,3743,3738
2,AGUDO,17533,17610,17682,17833,17917,18000,17114,17063,16666,16612,17161,17140,17120,17102,17085,16524,16461,16401,16344
3,AJURICABA,7660,7627,7588,7508,7463,7419,7411,7374,7221,7187,7420,7403,7387,7372,7358,7063,7024,6987,6951
4,ALECRIM,8287,8144,7984,7650,7465,7281,7380,7259,6935,6828,7010,6935,6865,6799,6736,6077,5950,5827,5710


<pre>Agora, aplique o pivoteamento do formato largo em formato longo, para que
    - o nome do município seja o índice das linhas;
    - uma coluna represente o atributo <b>ano</b>; e
    - outra coluna represente o atributo <b>Pop. Estimada</b>.</pre>

<pre>Isto é, modificaremos o DataFrame <b>df_populacao</b> do formato largo:</pre>

<table>
    <tr>
        <th>Município</th>
        <th>2001</th>
        <th>2002</th>
        <th>2003</th>
        <th>2004</th>
        <th>...</th>
        <th>2018</th>
        <th>2019</th>
        <th>2020</th>
        <th>2021</th>
    </tr>
    <tr>
        <td>ACEGUA</td>
        <td>3970</td>
        <td>4002</td>
        <td>4036</td>
        <td>4109</td>
        <td>...</td>
        <td>4858</td>
        <td>4901</td>
        <td>4942</td>
        <td>4981</td>
    </tr>
    <tr>
        <td>AGUA SANTA</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
        <td>...</td>
    </tr>
</table>

<pre>Para o formato longo:</pre>

<table>
    <tr>
        <th>Município</th>
        <th>ano</th>
        <th>Pop. Estimada</th>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2001</td>
        <td>3970</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2002</td>
        <td>4002</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2003</td>
        <td>4036</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2004</td>
        <td>4109</td>
    </tr>
    <tr>
        <th>...</th>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2018</td>
        <td>4858</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2019</td>
        <td>4901</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2020</td>
        <td>4942</td>
    </tr>
    <tr>
        <th>ACEGUA</th>
        <td>2021</td>
        <td>4981</td>
    </tr>
    <tr>
        <th>AGUA SANTA</th>
        <td>...</td>
        <td>...</td>
    </tr>
</table>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_populacao = df_populacao.melt(id_vars = 'Município', var_name = 'ano', value_name = 'Pop. Estimada')
```

<br/>

</details>

<pre>Agora, filtre os dados de população estimada do DataFrame <b>df_populacao</b> para que seja listada apenas a população estimada do ano mais recente.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_populacao = df_populacao[df_populacao['ano'] == df_populacao['ano'].max()]
```

<br/>

</details>

<pre>Uma vez que estamos usando apenas a população estimada do ano mais recente, não temos mais a necessidade de armazenar a variável <b>ano</b>.</pre>

<pre>Portanto, exclua a coluna <b>ano</b> do DataFrame <b>df_populacao</b>.</pre.>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_populacao = df_populacao.drop(columns='ano')
```

<br/>

</details>

## Junção, Mescla e Concatenação 🥇

### Junção

<pre>Vamos calcular a taxa de homicídios por habitante para cada um dos municípios do Rio Grande do Sul.</pre>

<pre>Para isso, será necessário juntar os DataFrames <b>df_homicidios</b> e <b>df_populacao</b>, utilizando um índice em comum como chave da junção.</pre>

<pre>No entanto, note que apenas o DataFrame <b>df_homicidios</b> utiliza o nome de Município como índice. Isto não ocorre para o DataFrame <b>df_populacao</b>.</pre>

<pre>Portanto, para que seja possível realizar a junção entre os DataFrames, defina a coluna <b>Município</b> como índice do DataFrame <b>df_populacao</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_populacao = df_populacao.set_index('Município')
```

<br/>

</details>

<pre>Agora, faça a junção dos DataFrames <b>df_homicidios</b> e <b>df_populacao</b>, utilizando o índice (em comum) como chave da junção, e atribua o resultado para o DataFrame <b>df_tx_de_homicidio</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_tx_de_homicidio = df_homicidios.join(df_populacao)
```

<br/>

</details>

<pre>Veja que o DataFrame <b>df_tx_de_homicidio</b> apresenta o número de homicídios dolosos por ano, bem como a totalização de cada município.</pre>

<pre>Como precisamos apenas da totalização de cada município, mantenha apenas as colunas <b>Total/Município</b> e <b>Pop. Estimada</b> no DataFrame <b>df_tx_de_homicidio</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_tx_de_homicidio = df_tx_de_homicidio[['Total/Município', 'Pop. Estimada']]
```

<br/>

</details>

<pre>Agora, crie a coluna <b>Taxa de Homicídio Doloso</b> no DataFrame <b>df_tx_de_homicidio</b>, obtida por operação de divisão entre <b>Total/Município</b> e <b>Pop. Estimada</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_tx_de_homicidio['Taxa de Homicídio Doloso'] = df_tx_de_homicidio['Total/Município'] / df_tx_de_homicidio['Pop. Estimada']
```

<br/>

</details>

<pre>Como a totalização de cada município e a população estimada não se fazem mais necessárias, mantenha apenas a coluna <b>Taxa de Homicídio Doloso</b> no DataFrame <b>df_tx_de_homicidio</b>.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_tx_de_homicidio = df_tx_de_homicidio[['Taxa de Homicídio Doloso']]
```

<br/>

</details>

<pre>Por fim, apresente todos os valores do DataFrame <b>df_tx_de_homicidio</b> com valores percentuais e precisão de duas casas decimais.</pre>

<details>
  <summary>Resposta</summary>

<br/>

```python
df_tx_de_homicidio.map(lambda x: f'{x:.2%}')
```

<br/>

</details>